pkg/security/podsecuritypolicy/seccomp/mustrunas.go

```diff
@@ -22,88 +22,62 @@ import (
 	annotationsPath := field.NewPath("pod", "metadata", "annotations")
 
-	if len(s.allowedProfiles) == 0 {
-		// if anything is set it is invalid.
-		return s.validateNoAnnotationsAreSet(pod, annotationsPath)
-	}
-
-	// at least pod should have something set and any container annotation
-	// must also be valid
-	return s.validateAnnotationsAreAllowed(pod, annotationsPath)
-}
-
-func (s *withSeccompProfile) validateAnnotationsAreAllowed(pod *api.Pod, fieldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
+	// set the allowed profils from either the values on the strategy or as the
+	// default profile is there are no values.
+	allowedProfiles := s.allowedProfiles
+	if allowedProfiles == nil || len(allowedProfiles) == 0 {
+		allowedProfiles = []string{s.defaultProfile}
+	}
+
 	podProfile, hasPodProfile := pod.Annotations[api.SeccompPodAnnotationKey]
 	// empty pod profile is allowed IF all the containers have their own valid annotation.
 	// If the container's do not have their own annotations we'll catch it in the logic
 	// below.
-	if hasPodProfile && !isProfileAllowed(podProfile, s.allowedProfiles) {
-		allErrs = append(allErrs, field.NotSupported(fieldPath.Child(api.SeccompPodAnnotationKey), podProfile, s.allowedProfiles))
+	if hasPodProfile && !isProfileAllowed(podProfile, allowedProfiles) {
+		allErrs = append(allErrs, field.NotSupported(annotationsPath.Child(api.SeccompPodAnnotationKey), podProfile, s.allowedProfiles))
 	}
 
 	for _, c := range pod.Spec.Containers {
 		containerProfileKey := api.SeccompContainerAnnotationKeyPrefix + c.Name
 		containerProfile := getProfileForContainer(pod, containerProfileKey, podProfile)
 
-		if !isProfileAllowed(containerProfile, s.allowedProfiles) {
-			allErrs = append(allErrs, field.NotSupported(fieldPath.Child(containerProfileKey), containerProfile, s.allowedProfiles))
-		}
-	}
-
-	return allErrs
-}
-
-// validateNoAnnotationsAreSet validates that seccomp annotations for the pod and containers
-// are not set.
-func (s *withSeccompProfile) validateNoAnnotationsAreSet(pod *api.Pod, fieldPath *field.Path) field.ErrorList {
-	allErrs := field.ErrorList{}
-	podProfile, hasPodProfile := pod.Annotations[api.SeccompPodAnnotationKey]
-
-	if hasPodProfile {
-		allErrs = append(allErrs, field.Invalid(fieldPath.Child(api.SeccompPodAnnotationKey), podProfile, "seccomp may not be set"))
-	}
-
-	for _, c := range pod.Spec.Containers {
-		containerProfileKey := api.SeccompContainerAnnotationKeyPrefix + c.Name
-		containerProfile, hasContainerProfile := pod.Annotations[containerProfileKey]
-
-		if hasContainerProfile {
-			allErrs = append(allErrs, field.Invalid(fieldPath.Child(containerProfileKey), containerProfile, "seccomp may not be set"))
+		if !isProfileAllowed(containerProfile, allowedProfiles) {
+			allErrs = append(allErrs, field.NotSupported(annotationsPath.Child(containerProfileKey), containerProfile, s.allowedProfiles))
```

Comment:
`field.Forbidden` would make more sense. In addition there should be a check for the profile string format. That could then return `field.NotSupported`.